Source/WebCore/dom/Document.cpp

```diff
@@ -3942,28 +3945,90 @@ ExceptionOr<void> Document::write(Document* entryDocument, SegmentedString&& tex
     return { };
 }
 
-ExceptionOr<void> Document::write(Document* entryDocument, FixedVector<String>&& strings)
+ExceptionOr<void> Document::write(Document* entryDocument, FixedVector<std::variant<RefPtr<TrustedHTML>, String>>&& strings)
 {
-    if (!isHTMLDocument() || m_throwOnDynamicMarkupInsertionCount)
-        return Exception { ExceptionCode::InvalidStateError };
+    auto isTrusted = true;
+    SegmentedString text;
+    for (auto& entry : strings) {
+        text.append(
+            WTF::switchOn(
+                WTFMove(entry),
+                [&isTrusted](const String& string) -> String {
+                    isTrusted = false;
+                    return string;
+                },
+                [](const RefPtr<TrustedHTML>& html) -> String {
+                    return html->toString();
+                }
+            )
+        );
+    }
+
+    if (isTrusted || !scriptExecutionContext()->settingsValues().trustedTypesEnabled)
+        return write(entryDocument, WTFMove(text));
+
+    String textString = String(text.toString());
+    auto stringValueHolder = trustedTypeCompliantString(TrustedType::TrustedHTML, *scriptExecutionContext(), textString, "Document write"_s);
+    if (stringValueHolder.hasException())
+        return stringValueHolder.releaseException();
+    text.clear();
+    text.close();
+    SegmentedString trustedSegmentedText(stringValueHolder.releaseReturnValue());
+    return write(entryDocument, WTFMove(trustedSegmentedText));
```

Comment:
These would read better without the local variable.

Reply:
In the new shared function I'm conditionally appending the linefeed to the end, is there a better way to do that without making a new SegmentedString variable?